File: src/libs/fs.js
var scoreName = scope.get("OBJECTIVE_NAME");
var tmpScoreName = "MoonCraftTmp";
var int = scope.get("int");

exports._fs_uint4 = function(fd, val, offsetX, offsetY, offsetZ, mode)
{
    var sel = exports.resolve_fd(fd);

    offsetX = "~" + (offsetX || "");
    offsetY = "~" + (offsetY || "");
    offsetZ = "~" + (offsetZ || "");

    var valSel = val.selector || val.name;
    var _scoreName = val.scoreName || scoreName;

    for(var i = 0; i < 16; i++)
    {
        command("execute {0} ~ ~ ~ detect {1} {2} {3} wool {4} scoreboard players set {5} {6} {4}"
            .format(sel, offsetX, offsetY, offsetZ, i, valSel, _scoreName));
    }
}

exports._fs_uint8 = function(fd, val, autoNewline)
{
    var sel = exports.resolve_fd(fd);
    var left = int();
    var right = int();

    exports._fs_uint4(sel, left);
    exports._fs_uint4(sel, right, 0, 0, 1);

    command("tp {0} ~1 ~ ~".format(sel));

    if(autoNewline)
        command("execute {0} ~ ~ ~ detect ~ ~ ~ air -1 tp @e[type=ArmorStand,c=1] 1 ~ ~2".format(sel));

    left.multiplicate(16);
    left.add(right);
    val.set(left);
}

exports.resolve_fd = function(fd)
{
    if(typeof fd == "string")
    {
        return fd;
    }
    else
    {
        var sel = "@e[type=ArmorStand,tag=fd]";
        var fdScoreName = fd.scoreName || scoreName;
        var selfSel = "@e[type=ArmorStand,c=1,r=0,tag=fd]";

        command("execute {0} ~ ~ ~ scoreboard players operation {1} {2} = {1} {3}".format(sel, selfSel, tmpScoreName, scoreName));
        command("execute {0} ~ ~ ~ scoreboard players operation {1} {2} -= {3} {4}".format(sel, selfSel, tmpScoreName, fd.name, scoreName));

        return "@e[type=ArmorStand,tag=fd,score_{0}_min=0,score_{0}=0]".format(tmpScoreName);
    }
}

File: src/libs/fs.lua
import("./fs.js")

-- fs at 1 10 5

function fs_init()
    command("kill @e[type=ArmorStand,tag=fs]")
    command("kill @e[type=ArmorStand,tag=file]")
    command("summon ArmorStand 1 10 5 {Tags:[\"fs\"],NoGravity:true}")

    while not "/execute @e[type=ArmorStand,tag=fs] ~ ~ ~ testforblock ~ ~ ~ air" do
        command("execute @e[type=ArmorStand,tag=fs] ~ ~ ~ summon ArmorStand ~ ~ ~ {Tags:[\"file\"],NoGravity:true}")
        command("tp @e[type=ArmorStand,tag=fs] ~ ~3 ~")
    end
end

local currFd = 0
function fopen(name)
    command("execute @e[type=ArmorStand,tag=file] ~ ~ ~ summon ArmorStand ~ ~ ~ {Tags:[\"fopen\"],NoGravity:true}")

    name[#name + 1] = 0

    for i = 1, #name do

        local selfScore = score("@e[type=ArmorStand,tag=fopen,c=1]", OBJECTIVE_NAME)
        _fs_uint8("@e[type=ArmorStand,tag=fopen]", selfScore)

        local curr = name[i]
        command("execute @e[type=ArmorStand,tag=fopen] ~ ~ ~ " ..
            "scoreboard players operation @e[type=ArmorStand,tag=fopen,c=1] "..OBJECTIVE_NAME.." -= "..js_eval("curr.name").." "..OBJECTIVE_NAME)

        command("kill @e[type=ArmorStand,tag=fopen,score_" .. OBJECTIVE_NAME .. "_min=1]")
        command("kill @e[type=ArmorStand,tag=fopen,score_" .. OBJECTIVE_NAME .. "=-1]")
    end

    local count = 0
    command("execute @e[type=ArmorStand,tag=fopen] ~ ~ ~ scoreboard players add " .. js_eval("count.name") .. " " .. OBJECTIVE_NAME .. " 1")

    if count == 1 then
        currFd = currFd + 1
        local fd = score("@e[type=ArmorStand,tag=fopen]", OBJECTIVE_NAME)
        fd = currFd
        command("tp @e[type=ArmorStand,tag=fopen] 1 ~ ~2")
        command("entitydata @e[type=ArmorStand,tag=fopen] {Tags:[\"fd\"]}")
        return currFd
    else
        return 0
    end
end

function fclose(fd)
    command("kill " + resolve_fd(fd))
end

function fgetc(fd)
    local val = 0
    _fs_uint8(fd, val, true)
    return val
end
